### PR TITLE
Navigate to nested router within stack

### LINKED ISSUE #704: Navigating to a nested screen
Supposing I have a StackNavigator nested within another StackNavigator like below, how would I navigate to the NewChatScreen from the root? Doing navigate("NewChat") doesn't seem to work.

```
const ChatNavigator = StackNavigator({
  List: { screen: ListScreen },
  NewChat: { screen: NewChatScreen }
});

const App = StackNavigator({
  Home: { screen: HomeScreen },
  Chat: { screen: ChatNavigator }
});

AppRegistry.registerComponent('App', () => App);
```

### LINKED ISSUE #983: How to Navigate Between Nested Stacks?
Let's say I have a main navigator:

```jsx
const Root = StackNavigator({
  Auth: { screen: AuthStack },
  Onboarding: { screen: OnboardingStack },
  Tabs: { screen: TabsStack },
});
```

And the following nested stacks:

```jsx
const AuthStack = StackNavigator({
  AuthOne: { screen: AuthOne },
  AuthTwo: { screen: AuthTwo },
  ...
});

const OnboardingStack = StackNavigator({
  OnboardingOne: { screen: OnboardingOne },
  OnboardingTwo: { screen: OnboardingTwo },
  ...
});

const TabsStack = TabNavigator({
  TabOne: { screen: TabOne },
  TabTwo: { screen: TabTwo },
  ...
});

```

Once I navigate from the main navigator into an child stack, say:

```jsx
navigation.navigate('Auth'); // navigates me to the AuthStack -> initialRoute = AuthOne
```

I seem to get "stuck" / "stranded" there. What I mean by that is I am now limited in the routes I can navigate to. The navigation system now seems to be only aware of `AuthOne` and `AuthTwo`. The following actions yield errors, where previously they would work:

```jsx
navigation.navigate('Onboarding'); // Error: There is no route defined for key Onboarding. Must be one of 'AuthOne','AuthTwo'
navigation.navigate('OnboardingOne'); // Error: There is no route defined for key OnboardingOne. Must be one of 'AuthOne','AuthTwo'
navigation.navigate('Tabs'); // Error: There is no route defined for key Tabs. Must be one of 'AuthOne','AuthTwo'
```

So my question is this: how do you navigate **between** nested / leaf stacks? Or worded a different way: how do you make each stack aware of all the routes?

### PR DESCRIPTION
**Pull request by grabbou**
Fixes #983 
Fixes #704

### CODE REVIEW

File: src/routers/__tests__/StackRouter-test.js [hunk 1]
```diff
@@ -283,6 +283,32 @@ describe('StackRouter', () => {
...
+      routes: [
+        { key: 'Init', routeName: 'foo' }
+      ]
+    });
+    const pushedState = TestRouter.getStateForAction(NavigationActions.navigate({ routeName: 'qux' }), initState);
+    // $FlowFixMe
+    expect(pushedState.index).toEqual(1);
+    // $FlowFixMe
+    expect(pushedState.routes[1].index).toEqual(1);
+    // $FlowFixMe
```
Comment (ericvicenti) [inline src/routers/__tests__/StackRouter-test.js]: hmm, I wonder what the best flow practices would be here

File: src/routers/StackRouter.js [hunk 2]
```diff
@@ -138,34 +138,67 @@ export default (
...
-            ...action,
+            params: action.params,
             key: _getUuid(),
             routeName: action.routeName,
           };
         }
         return StateUtils.push(state, route);
       }
 
+      // Handle navigation to other child routers that are not yet pushed
```
Comment (ericvicenti) [inline src/routers/StackRouter.js]: Awesome! I thought this was supported with the initial release, but its great to see explicit coverage of a very highly-requested feature

File: src/routers/__tests__/StackRouter-test.js [hunk 3]
```diff
@@ -283,6 +283,32 @@ describe('StackRouter', () => {
     });
   });
 
+  test('Pushes other navigators when navigating to an unopened route name', () => {
```
Comment (ericvicenti) [inline src/routers/__tests__/StackRouter-test.js]: This test makes it super easy to see what changed!

File: examples/NavigationPlayground/js/ModalStack.js [hunk 4]
```diff
@@ -55,6 +55,18 @@ MyProfileScreen.navigationOptions = ({ navigation }) => ({
   title: `${navigation.state.params.name}'s Profile!`,
 });
 
+const ProfileNavigator = StackNavigator({
+  Home: {
+    screen: MyHomeScreen,
+  },
+  Profile: {
```
Comment (ericvicenti) [inline examples/NavigationPlayground/js/ModalStack.js]: Ah. So by adding this extra layer of navigator, you were able to expose the bug

**Comment by richardfickling** [comment]
Any chance we could get a new npm release with this?

**Comment by henrikra** [comment]
So you can't still `reset` to nested navigation stack?

**Comment by malashkevich** [comment]
Hello.
Seems like this will fix my issue. When this fix will be available in npm release?

**Comment by gregblass** [comment]
Is this released yet or still sitting around?

**Comment by adkl** [comment]
I still have the same problem too.

**Comment by smithaitufe** [comment]
I really appreciate the work you guys are doing. However, this issue is far from fixed. It is very important for us.

**Comment by supercamilo** [comment]
This is still failing for me when my nested navigators are more than 2 levels deep. I can navigate to the first and second level but the third level is not handled correctly by `StackRouter.getStateForAction()`

**Comment by shicholas** [comment]
I'm running into the same problem as @supercamilo. This works fine for one level of nesting, but not more than one. My flow is:

MainNav (splash screen) => Drawer Nav => Stack Nav. Transitioning in the latest one does not work.

**Comment by supercamilo** [comment]
My workaround for my 3 levels deep navigators (with Redux) is handling navigation manually on the reducer. Like this:

```
const navigatorReducer = (state = initialNavState, action) => {
    let nextState;

    if (action.type.toString().startsWith('Navigation/')) {
        const mainState = MainNavigator.router.getStateForAction(action, state);

        switch (action.routeName) {
            // Manually get actions util react-navigation supports more than 2 levels deep nested StackNavigators
            case 'SignedIn': {
                const subState = SignedInNavigator.router.getStateForAction(action.action, state);
                mainState.routes[mainState.index].routes = [subState.routes[subState.index]];
                nextState = mainState;
                break;
            }
            case 'SignedOut': {
                const subState = SignedOutNavigator.router.getStateForAction(action.action, state);
                mainState.routes[mainState.index].routes = [subState.routes[subState.index]];
                nextState = mainState;
                break;
            }
            default:
                nextState = mainState;
                break;
        }
    } else {
        // All other custom actions
    }

    return nextState || state;
};
```

**Comment by morfsys** [comment]
Was the fix pushed? Still experiencing the same issue.

**Comment by brentvatne** [comment]
if you're still having this issue, please create a new issue and follow the provided issue template, happy to look at it! :)

**Pull request merged**